### DISCUSSION TITLE
mypy type error

### DISCUSSION THREAD

**Discussion by minyoung90** [sqlalchemy/sqlalchemy#9413] [Usage Questions] [answered]
### Describe the bug

I got mypy type error, but I don't understand.
is there something I was wrong?

### SQLAlchemy Version in Use

2.0.4

### DBAPI (i.e. the database driver)

pymysql

### Database Vendor and Major Version

MySQL 8.0.28

### Python Version

3.9.16

### Operating system

Ubuntu 20.04

### To Reproduce

```python
from typing import TypeVar, TYPE_CHECKING
from datetime import datetime

from sqlalchemy import inspect, String, DateTime, BigInteger
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.orm.decl_api import DeclarativeMeta, registry

Child = TypeVar("Child", bound="Base")

mapper_registry = registry()


class Base(metaclass=DeclarativeMeta):
    __abstract__ = True
    __name__: str
    __table_args__ = {"mysql_charset": "utf8mb4"}

    registry = mapper_registry
    metadata = mapper_registry.metadata

    def __init__(self, **kwargs):
        mapper = inspect(self).mapper
        for key in mapper.relationships:
            if key in kwargs:
                kwargs[key] = mapper.relationships[key].entity.class_(**kwargs[key])
        super().__init__(**kwargs)


class LoginHistory(Base):
    id: Mapped[int] = mapped_column(
        BigInteger, primary_key=True, autoincrement=True, comment="PK"
    )
    login_ip: Mapped[str] = mapped_column(String(45), nullable=False)
    login_time: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    if TYPE_CHECKING:
        _mypy_mapped_attrs = [
            id,
            login_ip,
            login_time,
        ]


login_history_obj = LoginHistory(
    login_ip="192.0.0.1",  # here, mypy gives me an error
    login_time=datetime.utcnow(),  # here, mypy gives me an error
)
```

### Error

```
Argument "login_ip" to "LoginHistory" has incompatible type "str"; expected "Mapped[str]"  [arg-type] mypy(error)
Argument "login_time" to "LoginHistory" has incompatible type "datetime"; expected "Mapped[datetime]"  [arg-type] mypy(error)
```

### Additional context

I installed mypy using "poetry add sqlalchemy[mypy]"

and in my project root, I added this config file (mypy.ini)
```
[mypy]
plugins = sqlalchemy.ext.mypy.plugin
```

**Comment by CaselIT** [comment]
Hi,

Can you try removing the plugin, the `_mypy_mapped_attrs` and creating base subclassing `DeclarativeBase`?

**Comment by minyoung90** [comment]
```python
from typing import TypeVar
from datetime import datetime

from sqlalchemy import inspect, String, DateTime, BigInteger
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.orm.decl_api import DeclarativeBase, registry

Child = TypeVar("Child", bound="Base")

mapper_registry = registry()


class Base(DeclarativeBase):
    __abstract__ = True
    __table_args__ = {"mysql_charset": "utf8mb4"}

    registry = mapper_registry
    metadata = mapper_registry.metadata

    def __init__(self, **kwargs):
        mapper = inspect(self).mapper
        for key in mapper.relationships:
            if key in kwargs:
                kwargs[key] = mapper.relationships[key].entity.class_(**kwargs[key])
        super().__init__(**kwargs)


class LoginHistory(Base):
    id: Mapped[int] = mapped_column(
        BigInteger, primary_key=True, autoincrement=True, comment="PK"
    )
    login_ip: Mapped[str] = mapped_column(String(45), nullable=False)
    login_time: Mapped[datetime] = mapped_column(DateTime, nullable=False)


login_history_obj = LoginHistory(
    login_ip="192.0.0.1",  # error is gone!
    login_time=datetime.utcnow(),  # error is gone!
)
```
I did and, the error is gone now! Thanks!
but, document has still a description about "_mypy_mapped_attrs" @CaselIT 
https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html

**Comment by CaselIT** [comment, marked as answer]
Yes, but I'm not sure when that helps.
The mypy extension is deprecated as mentioned in that page, things should be better without the plugin, and other type checkers mostly work too

**Comment by claudiosw** [comment]
Doing like this:
profession_id: Mapped[uuid.UUID] = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
I was getting:
error: Incompatible types in assignment (expression has type "Column[UUID]", variable has type "Mapped[UUID]")
Analysing this solution above, I tried to use mapped_column instead of Column and it worked.
But this doc seemed that the original solution should work.
https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html
Not sure what was the problem.

**Reply by CaselIT** [reply]
The advice is to disable the plugin and use mapped_columns

**Comment by behnazh-w** [comment]
While the types work fine using `DeclarativeBase` and `mapped_column`,  I get  `incompatible type "str"; expected "Dict[str, Any]"` error using a `@declarative_mixin`

```python
@declarative_mixin
class A:
    description: Mapped[str] = mapped_column()

class B(DeclarativeBase):
    ...

class C(A, B):
    ...

C(description="foo")  # mypy error
```

error:
```bash
 error: Argument "description" to "C" has incompatible type "str"; expected "Dict[str, Any]"  [arg-type]

```

**Reply by claudiosw** [reply]
Here is my code, it may help you:

```
import uuid
from sqlalchemy import String
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import UUID
from src.infra.db_models.db_base import Base


class ProfessionsDBModel(Base):
    """ Defines the professions database model.
    """

    __tablename__ = "professions"

    profession_id: Mapped[uuid.UUID] = \
        mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(80), nullable=False, unique=True)
    description: Mapped[str] = mapped_column(String(200), nullable=False)
```

**Reply by behnazh-w** [reply]
Thanks @claudiosw . My issue is with `@declarative_mixin` though. Your example doesn't seem to be using it.

**Reply by claudiosw** [reply]
Which version SQLAlchemy are you using?
In [this link](https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html) is written:
Changed in version 2.0: For users coming from the 1.4 series of SQLAlchemy who may have been using the [mypy plugin](https://docs.sqlalchemy.org/en/20/orm/extensions/mypy.html), the [declarative_mixin()](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.declarative_mixin) class decorator is no longer needed to mark declarative mixins, assuming the mypy plugin is no longer in use.

**Reply by behnazh-w** [reply]
I'm using SQLAlchemy version 2.0.15 but didn't know I don't need to use `declarative_mixin` anymore, thanks.

After removing the `declarative_mixin`  I noticed `mypy` has other issues with handling inheritance in my case. Thanks for the help anyway!

**Reply by claudiosw** [reply]
You're welcome!
